In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
#Setting plot sizes and type of plot
plt.rc("font", size=14)
plt.rcParams['axes.grid'] = True
plt.figure(figsize=(6,3))
plt.gray()
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn import metrics
from sklearn.preprocessing import StandardScaler, MinMaxScaler, MaxAbsScaler

<Figure size 432x216 with 0 Axes>

In [4]:
from sklearn.linear_model import LogisticRegression, LinearRegression, ElasticNet, Lasso, Ridge
from sklearn.neighbors import KNeighborsClassifier, KNeighborsRegressor
# from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor, export_graphviz, export
from sklearn.ensemble import BaggingClassifier, BaggingRegressor,RandomForestClassifier,RandomForestRegressor
from sklearn.ensemble import GradientBoostingClassifier,GradientBoostingRegressor, AdaBoostClassifier, AdaBoostRegressor 
from sklearn.svm import LinearSVC, LinearSVR, SVC, SVR
# from xgboost import XGBClassifier
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix

In [5]:
data=pd.read_csv('C:/Users/hp/Desktop/Python projects/ML projects/archive.zip')

In [12]:
# Checking the transaction distribution
Total_transactions= len(data)
normal=len(data[data.Class==0])
fraudlent= len(data[data.Class==1])
fraud_percentage= round(fraudlent/normal *100,2)
print('Total number of transactions are {}' .format(Total_transactions))
print('Number of normal transactions are {}' .format(normal))
print('Number of fraudlent transactions are {}' .format(fraudlent))
print('Percentage of fraud transactions are {}'.format(fraud_percentage))

Total number of transactions are 284807
Number of normal transactions are 284315
Number of fraudlent transactions are 492
Percentage of fraud transactions are 0.17


In [13]:
# We can also check for null values
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [14]:
min(data.Amount),max(data.Amount)

(0.0, 25691.16)

In [15]:
# it is a good practice to scale this variable. We can use a standard scaler to make it fix.
sc=StandardScaler()
amount=data['Amount'].values
data['Amount']=sc.fit_transform(amount.reshape(-1,1))

In [16]:
data.drop(['Time'],axis=1,inplace=True)

In [18]:
data.head()

V1        V2        V3        V4        V5        V6        V7  \
0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9       V10  ...       V21       V22       V23       V24  \
0  0.098698  0.363787  0.090794  ... -0.018307  0.277838 -0.110474  0.066928   
1  0.085102 -0.255425 -0.166974  ... -0.225775 -0.638672  0.101288 -0.339846   
2  0.247676 -1.514654  0.207643  ...  0.247998  0.771679  0.909412 -0.689281   
3  0.377436 -1.387024 -0.054952  ... -0.108300  0.005274 -0.190321 -1.175575   
4 -0.270533  0.817739  0.753074  ... -0.009431  0.798278 -0.137458  0.141267   

        V25       V26       V27       V28    Amount  Class  
0  0.128539 -0.189115  0.133558 -0.021053  0.244964      0  
1  0.167170  0.125895 -0.008983  0.014724 -0.342475      0  
2 -0.327642 -0.139097 -0.055353 -0.059752  1.160686      0  
3  0.647376 -0.221929  0.062723  0.061458  0.140534      0  
4 -0.206010  0.502292  0.219422  0.215153 -0.073403      0  

[5 rows x 30 columns]

In [20]:
data.shape

(284807, 30)

In [21]:
data.drop_duplicates(inplace=True)

In [24]:
data.shape # after removing duplicate transactions

(275663, 30)

In [25]:
# Before splitting train & test — we need to define dependent and independent variables. The dependent variable is also 
# known as X and the independent variable is known as y.
X = data.drop('Class', axis = 1).values
y = data['Class'].values

In [27]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 1)
# We now have two different data set — Train data we will be used for training our model and the data which is unseen will be used for testing.

# Decision Tree

In [28]:
DT =DecisionTreeClassifier(max_depth=4,criterion='entropy')
DT.fit(X_train,y_train)
dt_result=DT.predict(X_test)

In [29]:
# Checking the accuracy of decision tree model
print('Accuracy score of decision tree model is {}' .format(accuracy_score(y_test,dt_result)))

Accuracy score of decision tree model is 0.9991583957281328


In [30]:
# Checking F1 score
print('F1 score of the Decision Tree model is {}' .format(f1_score(y_test,dt_result)))

F1 score of the Decision Tree model is 0.7521367521367521


In [31]:
# Checking the confusion matrix
confusion_matrix(y_test,dt_result,labels=[0,1])

array([[68770,    18],
       [   40,    88]], dtype=int64)

First row represents positive and the second row represents negative. So, we have 68770 as true positive and 18 are false positive. That says, out of 68770+18=68788, we have 68770 that are successfully classified as a normal transaction and 18 were falsely classified as normal — but they were fraudulent.

# K-Nearest Neighbors

In [36]:
n = 7
KNN = KNeighborsClassifier(n_neighbors = n)
KNN.fit(X_train, y_train)
knn_result = KNN.predict(X_test)

In [37]:
# checking the accuracy of our K-Nearest Neighbors model
print('Accuracy score of the K-Nearest Neighbors model is {}'.format(accuracy_score(y_test, knn_result)))

Accuracy score of the K-Nearest Neighbors model is 0.999288989494457


In [38]:
# Checking the F1 score
print('F1 score of the K-Nearest Neighbors model is {}'.format(f1_score(y_test, knn_result)))

F1 score of the K-Nearest Neighbors model is 0.7949790794979079


# Logistic Regression

In [39]:
lr = LogisticRegression()
lr.fit(X_train, y_train)
lr_result = lr.predict(X_test)

In [40]:
# check the accuracy of our Logistic Regression model
print('Accuracy score of the Logistic Regression model is {}'.format(accuracy_score(y_test, lr_result)))

Accuracy score of the Logistic Regression model is 0.9989552498694062


In [41]:
# Checking F1-Score for the Logistic Regression model
print('F1 score of the Logistic Regression model is {}'.format(f1_score(y_test, lr_result)))

F1 score of the Logistic Regression model is 0.6666666666666666


# Support Vector Machines

In [42]:
svm = SVC()
svm.fit(X_train, y_train)
svm_result = svm.predict(X_test)

In [43]:
# checking the accuracy of Support Vector Machines model
print('Accuracy score of the Support Vector Machines model is {}'.format(accuracy_score(y_test, svm_result)))

Accuracy score of the Support Vector Machines model is 0.999318010331418


In [44]:
# Checking F1-Score for the Support Vector Machines model
print('F1 score of the Support Vector Machines model is {}'.format(f1_score(y_test, svm_result)))

F1 score of the Support Vector Machines model is 0.7813953488372093


# Random Forest

In [46]:
rf = RandomForestClassifier(max_depth = 4)
rf.fit(X_train, y_train)
rf_result = rf.predict(X_test)

In [47]:
# checking the accuracy of Random Forest model
print('Accuracy score of the Random Forest model is {}'.format(accuracy_score(y_test, rf_result)))

Accuracy score of the Random Forest model is 0.9991293748911718


In [48]:
# Checking F1-Score for the Random Forest model
print('F1 score of the Random Forest model is {}'.format(f1_score(y_test, rf_result)))

F1 score of the Random Forest model is 0.7222222222222223


 We just received 99.93% accuracy in our credit card fraud detection. This number should not be surprising as our data was balanced towards one class. The good thing that we have noticed from the confusion matrix is that — our model is not overfitted.